File: go/clients/messaging/decorators/eventstack.go
package decorators

import (
	"context"
	"time"

	oteltrace "go.opentelemetry.io/otel/trace"

	clientstack "github.com/gt-tech-ai/knowledge-engine/go/clients/decorators"
	"github.com/gt-tech-ai/knowledge-engine/go/core/interfaces"
	"github.com/gt-tech-ai/knowledge-engine/go/foundation/resilience/deadletter"
)

// EventStackDeps carries the collaborators for the §6.3 EventHandler stack. A nil layer
// (or zero timeout) is skipped, so callers opt into exactly the concerns they have wired.
type EventStackDeps struct {
	// Dedup skips a message whose key was already processed (nil disables dedup).
	Dedup interfaces.Deduplicator

	// KeyOf extracts the dedup key from a message; nil defaults to msg.ID.
	KeyOf func(*interfaces.Message) string

	// DeadLetter routes a message that still fails after retries (nil disables the DLQ).
	DeadLetter *deadletter.Queue

	// Retrier retries a transient handler failure (nil disables retry).
	Retrier interfaces.Retrier

	// CircuitBreaker fails fast when the handler's downstream is down (nil disables it).
	CircuitBreaker interfaces.CircuitBreaker

	// Tracer emits a span per handled message (nil disables tracing).
	Tracer oteltrace.Tracer

	// Metrics records per-message counts/errors/latency (nil disables metrics).
	Metrics interfaces.Metrics

	// Logger records a failed handle at Debug (nil disables the log).
	Logger interfaces.Logger

	// Name labels the handler in metrics and span names.
	Name string

	// Timeout bounds each handle attempt; zero disables the timeout layer.
	Timeout time.Duration
}

// WrapHandler composes the charter §6.3 EventHandler stack around inner, outermost →
// innermost: Dedup → DeadLetter → [Timeout → Retry → CircuitBreaker → Tracing → Metrics →
// Logging] → Handler. (Decode is the handler's own concern.) The inner resilience +
// observability reuses the shared client stack, so the mechanism is composed once, not
// re-implemented (charter §6.3). A message that still fails after retries is routed to the
// dead-letter queue and acked when it lands there; pair Dedup with a DeadLetter so a
// terminal failure is dead-lettered rather than skipped as a duplicate on redrive.
func WrapHandler(
	inner interfaces.MessageHandler,
	deps EventStackDeps,
) interfaces.MessageHandler {
	keyOf := deps.KeyOf
	if keyOf == nil {
		keyOf = func(m *interfaces.Message) string { return m.ID }
	}
	stack := clientstack.New(deps.Name).
		WithRetrier(deps.Retrier).
		WithCircuitBreaker(deps.CircuitBreaker).
		WithTimeout(deps.Timeout).
		WithTracer(deps.Tracer).
		WithMetrics(deps.Metrics).
		WithLogger(deps.Logger)

	return func(ctx context.Context, msg *interfaces.Message) error {
		// Dedup (outermost): a message already processed is acked without re-running.
		if deps.Dedup != nil {
			if seen, err := deps.Dedup.Seen(ctx, keyOf(msg)); err == nil && seen {
				return nil
			}
		}
		// Inner: resilience + observability via the shared client stack.
		_, err := clientstack.Run(
			ctx,
			stack,
			"handle",
			clientstack.RunOpts{Retryable: true},
			func(c context.Context) (struct{}, error) { return struct{}{}, inner(c, msg) },
		)
		if err == nil {
			return nil
		}
		// DeadLetter: after retries + CB exhaust, route the poison message; ack when it
		// lands, otherwise return the error so the source message is redriven.
		if deps.DeadLetter != nil && deps.DeadLetter.Send(ctx, interfaces.DeadLetter{
			ID:      msg.ID,
			Payload: msg.Payload,
			Reason:  err.Error(),
		}) {
			return nil
		}
		return err
	}
}

File: python/techai_webutils/src/techai_webutils/clients/decorators/event_stack.py
"""EventHandler decorator stack for a message handler coroutine (ARCHITECTURE.md#decorator-order).

Composes, outermost -> innermost, ``Dedup -> DeadLetter -> [Retry -> CircuitBreaker -> Timeout]
-> Handler``. Mirrors Go's ``clients/messaging/decorators.WrapHandler``. A message already
processed is skipped; one that still fails after retries is routed to the dead-letter queue and
acked when it lands there. Built from the existing foundation
primitives (``retry_transient_async`` + ``with_timeout`` + the circuit breaker) and the DLQ
facade, so the mechanism is composed once, not re-implemented.

Pair ``dedup`` with a ``dlq`` so a terminal failure is dead-lettered rather than skipped as a
duplicate on redrive.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING

from techai_webutils.core.interfaces.dlq import DeadLetter
from techai_webutils.foundation.resilience.async_retry import retry_transient_async
from techai_webutils.foundation.resilience.timeout import with_timeout

if TYPE_CHECKING:
    from collections.abc import Awaitable, Callable

    from techai_webutils.core.interfaces.circuit_breaker import CircuitBreakerInterface
    from techai_webutils.core.interfaces.dedup import Deduplicator
    from techai_webutils.core.interfaces.messaging import Message, MessageHandler
    from techai_webutils.foundation.resilience.dlq import DeadLetterQueue


@dataclass(frozen=True, slots=True)
class EventStackDeps:
    """Collaborators for the EventHandler stack.

    A None layer (or non-positive timeout / ``retry_max_attempts`` <= 1) is skipped, so callers
    opt into exactly the concerns they have wired.
    """

    dedup: Deduplicator | None = None
    """Skips a message whose key was already processed (None disables dedup)."""

    key_of: Callable[[Message], str] | None = None
    """Extracts the dedup key from a message; None defaults to ``msg.id``."""

    dlq: DeadLetterQueue | None = None
    """Routes a message that still fails after retries (None disables the DLQ)."""

    circuit_breaker: CircuitBreakerInterface | None = None
    """Fails fast when the handler's downstream is down (None disables it)."""

    retry_max_attempts: int = 1
    """Retries a transient handler failure; <= 1 disables retry."""

    timeout_seconds: float | None = None
    """Bounds each handle attempt, in seconds; None/<= 0 disables the timeout layer."""


def wrap_handler(inner: MessageHandler, deps: EventStackDeps) -> MessageHandler:
    """Compose the EventHandler stack around ``inner``.

    Outermost -> innermost: ``Dedup -> DeadLetter -> [Retry -> CircuitBreaker -> Timeout] ->
    Handler``. A message whose key was already processed is acked without re-running; one that
    still fails after retries is routed to the dead-letter queue and acked when it lands there,
    otherwise the failure is re-raised so the source message is redriven.
    """
    key_of: Callable[[Message], str] = deps.key_of or (lambda m: m.id)

    async def attempt(msg: Message) -> None:
        """Run one handle attempt under the optional timeout + circuit breaker."""
        run: Awaitable[None] = (
            with_timeout(inner(msg), deps.timeout_seconds)
            if deps.timeout_seconds and deps.timeout_seconds > 0
            else inner(msg)
        )
        if deps.circuit_breaker is not None:
            with deps.circuit_breaker:
                await run
        else:
            await run

    runner: MessageHandler = attempt
    if deps.retry_max_attempts > 1:
        runner = retry_transient_async(max_attempts=deps.retry_max_attempts)(attempt)

    async def wrapped(msg: Message) -> None:
        """Dedup-gate, run the retry/timeout/CB inner, and dead-letter a terminal failure."""
        if deps.dedup is not None and await deps.dedup.seen(key_of(msg)):
            return
        try:
            await runner(msg)
        except Exception as exc:  # terminal failure is routed to the DLQ (or redriven)
            if deps.dlq is not None and await deps.dlq.send(
                DeadLetter(id=msg.id, payload=msg.payload, reason=str(exc)),
            ):
                return
            raise

    return wrapped
